File: 139-单词拆分.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.size();
        vector<bool> nums(n, false);

        for(int i = n - 1; i > -1; i--){
            for(auto word : wordDict){
                if(i + word.size() <= n && string(s.substr(i, word.size())) == word){
                    if(i + word.size() == n || nums[i + word.size()])
                        nums[i] = true;
                }
            }
        }
        
        return nums[0];
    }
};

File: 139-单词拆分.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = [False] * len(s) + [True]
        word_set = set(wordDict)
        for i in range(len(s) - 1, -1, -1):
            for j in range(i, len(s) + 1):
                if s[i:j] in word_set and dp[j]:
                    dp[i] = True
        return dp[0]
